inc/Integrations/SalesforceD2C/Auth/SalesforceD2CAuth.php

```diff
@@ -0,0 +1,118 @@
+<?php declare(strict_types = 1);
+
+namespace RemoteDataBlocks\Integrations\SalesforceD2C\Auth;
+
+use WP_Error;
+
+/**
+ * Salesforce D2C Auth class.
+ *
+ * This class is used to authenticate with Salesforce D2C using a client ID and secret.
+ */
+class SalesforceD2CAuth {
+
+	/**
+	 * Generate a token from a client ID and secret, or use an existing token if available.
+	 *
+	 * @param string $endpoint The endpoint prefix URL for the data source,
+	 * @param string $organization_id The organization ID for the data source.
+	 * @param string $client_id The client ID (a version 4 UUID).
+	 * @param string $client_secret The client secret.
+	 * @return string|WP_Error The token or an error.
+	 */
+	public static function generate_token(
+		string $endpoint,
+		string $client_id,
+		string $client_secret
+	): string|WP_Error {
+		$saved_access_token = self::get_saved_access_token( $client_id );
+
+		if ( null !== $saved_access_token ) {
+			return $saved_access_token;
+		}
+
+		$access_token = self::get_token_using_client_credentials( $client_id, $client_secret, $endpoint );
+		return $access_token;
```

Comment:
FYI we can use nullish coalescing for brevity
```suggestion
		return self::get_saved_access_token( $client_id ) ?? self::get_token_using_client_credentials( $client_id, $client_secret, $endpoint );
```

Reply:
I used to prefer this but I remember hearing in a lot of reviews that people found it hard to follow in terms of the logic. So I ended up trying to strike a balance as to when one is used over the other. 

That said, in this case happy to switch to using null coalescing as it's honestly a lot shorter.